daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueHasherSpec.scala

```diff
@@ -0,0 +1,268 @@
+    builder += None -> ValueText("abcd-äöü€")
+    builder += None -> ValueParty(Party.assertFromString("Alice"))
+    builder += None -> ValueUnit
+    builder += None -> ValueNone
+    builder += None -> ValueOptional(Some(ValueText("Some")))
+    builder += None -> ValueList(FrontStack(ValueText("A"), ValueText("B"), ValueText("C")))
+    builder += None -> ValueVariant(None, Name.assertFromString("Variant"), ValueInt64(0))
+    builder += None -> ValueRecord(
+      None,
+      ImmArray(
+        None -> ValueText("field1"),
+        None -> ValueText("field2")
+      ))
+    builder += None -> ValueTextMap(
+      SortedLookupList(
+        Map(
+          "keyA" -> ValueText("valueA"),
+          "keyB" -> ValueText("valueB")
+        )))
+    val fields = builder.result()
+
+    ValueRecord(None, fields)
+  }
+
+  "hashValue" should {
+
+    import ValueHasher._
+
+    "be stable" in {
+      val tmpId = templateId("module", "name")
+      hashValueString(complexValue) shouldBe "b83cca003b2f90ea4ce9986c5ad1d654d6e2b345f4665474c1567d154412942a"
+      hashContractString(tmpId, complexValue) shouldBe "07e7b5534931dfca8e1b485c105bae4e10808bd13ddc8e897f258015f9d921c5"
+      hashContractKeyString(tmpId, complexValue) shouldBe "59b59ad7a6b6066e77b91ced54b8282f0e24e7089944685cb8f22f32fcbc4e1b"
+    }
+
+    "be deterministic and thread safe" in {
+      // Compute many hashes in parallel, check that they are all equal
+      // Note: intentionally does not reuse value instances
+      val hashes = Vector.fill(1000)(complexValue).par.map(hashValueString)
+      hashes.toSet.size shouldBe 1
+    }
+
+    "not produce collision in template id" in {
+      // Same value but different template ID should produce a different hash
+      val value = ValueText("A")
+
+      val hashes = Ref.LedgerString.Array(
+        hashContractString(templateId("AA", "A"), value),
+        hashContractString(templateId("A", "AA"), value),
+        hashContractKeyString(templateId("AA", "A"), value),
+        hashContractKeyString(templateId("A", "AA"), value),
+      )
+
+      for {
+        i <- hashes.indices
+        j <- hashes.indices
+        if (i < j)
+      } hashes(i) shouldNot be(hashes(j))
+
+    }
+
+    // Note: value version is given by the template ID, this check is not necessary
+    /*
+    "not produce collision in value version" in {
+      // Same value but different value version should produce a different hash
+      val value1 = ValueText("A")
+      val value2 = ValueText("A")
+
+      val tid = templateId("module", "name")
+
+      val hash1 = KeyHasher.hashValueString(value1, tid)
+      val hash2 = KeyHasher.hashValueString(value2, tid)
+
+      hash1.equals(hash2) shouldBe false
+    }
+     */
+
+    "not produce collision in list of text" in {
+      // Testing whether strings are delimited: ["AA", "A"] vs ["A", "AA"]
+      val value1 =
+        ValueList(FrontStack(ValueText("AA"), ValueText("A")))
```

Comment:
```suggestion
        VA.list(VA.text).inj(Vector("AA", "A"))
      // given import TypedValueGenerators.{ValueAddend => VA}
```

`ValueAddend` can variously be used to make hand-written (well-typed) `Value`s more obvious in tests. Most of the other collision tests in this file can benefit.

Reply:
I am not sure how to use this. 
Your example does not type check.

Reply:
I replace some code with `VA.list` and `VA.map`.
Is there other place where I can do similar refactoring ?

Reply:
I rewrite the whole test with `ValueAddend`